Add BigQuery profiler (#2445)
databrickslabs/remorph: docs/lakebridge/docs/assessment/profiler/bigquery.mdx

---
sidebar_position: 3
title: BigQuery Profiler Details
---
import Admonition from '@theme/Admonition';

# BigQuery Profiler Details

- [Prerequisites](#prerequisites)
- [Configure Connection to BigQuery](#configure-connection-to-bigquery)
- [Execute the profiler](#execute-the-profiler)
- [Pricing and cost projection](#pricing-and-cost-projection)
- [Generate the dashboard](#generate-the-dashboard)

## Prerequisites

### 1. Download
- gcloud CLI (https://cloud.google.com/sdk/docs/install)

> 
> The BigQuery profiler uses the `google-cloud-bigquery` Python client (installed automatically by the pipeline runner); no ODBC driver or platform-specific binary is required.

Reviewer: ```suggestion
- No driver install required for the BigQuery profiler
```

### 2. Authenticate to GCP

The profiler authenticates with [Application Default Credentials (ADC)](https://docs.cloud.google.com/docs/authentication/application-default-credentials). The simplest setup is:

```bash
gcloud auth application-default login
```

> For non-interactive use, ADC also supports a service-account JSON key (via `GOOGLE_APPLICATION_CREDENTIALS`) or [service-account impersonation](https://docs.cloud.google.com/docs/authentication/set-up-adc-local-dev-environment#sa-impersonation) — whichever your environment is configured for.

Reviewer: ```suggestion
For more options, please review the docs.
```

### Percentile families

`bq_slots_pricing_analysis` computes **six cost percentiles** per (project, region, time_window, workload_type):

| Percentile | Meaning |
|---|---|
| `avg` | Average slot demand over the window |
| `50th` | Median slot demand |
| `90th` | 90th-percentile burst capacity |
| `99th` | 99th-percentile burst capacity |
| `max` | Peak slot demand |
| `perf_based` | Derived directly from BigQuery's billed `slot_secs` divided by the `db_etl_effective_price_perf` factor — a "what it would have cost on Databricks at observed efficiency" estimate |

The dashboard's "Estimation approach" filter widget lets you choose which percentile to project. **The default for `monthly_weight_projection` in `TUNING_INPUT_PARAMS` is `db_price_90th`** — a balanced choice between average and peak.

### What the dashboard shows

- BigQuery vs Databricks cost comparison by month, grouped by `workload_type` (BI vs ETL).
- Per-project-region cost breakdown.
- Table-storage migration sizing (`TB_to_migrate`, archival storage costs on the target cloud).
- Reservation / commitment cost flow (when the customer is on BigQuery Editions and `exclude_reservations_data: false`).

> ## Generate the dashboard

Reviewer: can we bump this section after execute and the pricing breakdown can be a subsection for interested users.

---
sidebar_position: 3
title: BigQuery Profiler Details
---
import Admonition from '@theme/Admonition';

# BigQuery Profiler Details

- [Prerequisites](#prerequisites)
- [Configure Connection to BigQuery](#configure-connection-to-bigquery)
- [Execute the profiler](#execute-the-profiler)
- [Pricing and cost projection](#pricing-and-cost-projection)
- [Generate the dashboard](#generate-the-dashboard)

## Prerequisites

### 1. Download
- gcloud CLI (https://cloud.google.com/sdk/docs/install)

The BigQuery profiler uses the `google-cloud-bigquery` Python client (installed automatically by the pipeline runner); no ODBC driver or platform-specific binary is required.

### 2. Authenticate to GCP using gcloud CLI

```bash
gcloud auth application-default login
```

The profiler uses Google's `google.auth.default()` chain, attempting authentication in this order:

1. **`GOOGLE_APPLICATION_CREDENTIALS`** environment variable pointing at a service-account JSON key.
2. **Application Default Credentials** (from `gcloud auth application-default login`).
3. **Compute Engine / Cloud Run / GKE metadata server** (only if you happen to be running the profiler on GCP).

> For production engagements, customers typically supply a service-account JSON key file path during `configure-database-profiler`. For local testing, ADC works.

Reviewer: using ADC is completely fine for production. no need to tell google internals

For least-privilege deployments, the discrete permissions are:

```
bigquery.jobs.create
bigquery.jobs.list
bigquery.datasets.get
bigquery.datasets.getIamPolicy
bigquery.tables.get
bigquery.tables.getData
bigquery.routines.list
bigquery.readSessions.create
bigquery.reservations.list                  # Editions customers only
bigquery.reservationAssignments.list        # Editions customers only
bigquery.capacityCommitments.list           # Editions customers only
```

Grant via `gcloud`:

```bash
gcloud projects add-iam-policy-binding <project_id> \
  --member="user:<email>" \
  --role="roles/bigquery.admin"
```

Or using a dedicated service account:

```bash
gcloud iam service-accounts create lakebridge-profiler \
  --description="Lakebridge BigQuery profiler" \
  --display-name="lakebridge-profiler"
gcloud projects add-iam-policy-binding <project_id> \
  --member="serviceAccount:lakebridge-profiler@<project_id>.iam.gserviceaccount.com" \
  --role="roles/bigquery.admin"
gcloud iam service-accounts keys create ~/lakebridge-profiler-key.json \
  --iam-account=lakebridge-profiler@<project_id>.iam.gserviceaccount.com

> ```

Reviewer: I find it enough to list the needed permssions and leave to end users how to apply them

| Streaming inserts | `region-{region}.INFORMATION_SCHEMA.STREAMING_TIMELINE_BY_PROJECT` | `bigquery.jobs.list` |
| Write API | `region-{region}.INFORMATION_SCHEMA.WRITE_API_TIMELINE_BY_PROJECT` | `bigquery.jobs.list` |
| Reservation timeline | `region-{region}.INFORMATION_SCHEMA.RESERVATIONS_TIMELINE`, `RESERVATION_CHANGES` | `bigquery.reservations.list` |
| Capacity commitments | `region-{region}.INFORMATION_SCHEMA.CAPACITY_COMMITMENT_CHANGES_BY_PROJECT` | `bigquery.capacityCommitments.list` (admin project only) |

:::warning Attention:
Skip the reservation/commitment permissions if the customer is on **Pay-as-you-go pricing** rather than BigQuery Editions. The corresponding SQL files will return empty results — harmless — and the dashboard widgets will just be empty. You can also set `exclude_reservations_data: true` during configuration to skip those queries entirely (saves BigQuery slot time; the corresponding DuckDB tables are still created as empty stubs so the dashboard's commitment/consumption widgets resolve cleanly).
:::

:::tip
Unlike Synapse, BigQuery uses OAuth bearer tokens via gRPC, so MFA does not interfere with the profiler.

> :::

Reviewer: not useful to customers

Redact query text in extracted data? (default: no):
Exclude reservations and commitments data? (default: no):
Exclude streaming and write API summary? (default: no):
Exclude pricing analysis (skip step 2)? (default: no):
```

The resulting `~/.databricks/labs/lakebridge/.credentials.yml` looks like:

```yaml
secret_vault_type: local
secret_vault_name: null
bigquery:
  projects:
    - customer-prod-1
    - customer-admin
  regions:
    - us
  service_account_key_path: null      # null = use ADC
  profiling_window_days: 180
  target_cloud: gcp
  max_parallel_sqls: 8
  profiler:
    redact_query_text: false
    exclude_reservations_data: false
    exclude_streaming_metrics: false
    exclude_pricing_analysis: false
```

## Execute the profiler

```bash
databricks labs lakebridge execute-database-profiler --source-tech bigquery
```

The pipeline runs two steps:

1. **`bq_metadata_extract`** — connects to BigQuery and runs 16 vendored SQL queries serially across `(project, region)` pairs and in parallel within each iteration (`max_parallel_sqls` workers). Results land in `~/.databricks/labs/lakebridge_profilers/bigquery_assessment/profiler_extract.db` (DuckDB).

> 2. **`compute_pricing_analysis`** — runs in-process via DuckDB on the tables loaded by step 1, producing per-percentile cost projections, monthly weighted pricing, and a `consumption_by_commitment` view.

Reviewer: shouldn't this be part of TCO app. I am not against profilers owning this logic but strategically, this is a secret recipe

Author: Keeping pricing in this PR. It matches the Synapse and MSSQL profiler precedent — both ship their cost-projection step alongside the metadata extract so the CLI flow produces an actionable dashboard in one run. happy to migrate later if the TCO-app conversation lands on a different home, but extracting it now would leave the profiler without an end-to-end.
